### ISSUE TITLE
Unable to import all Shaarli bookmarks in Firefox

### ISSUE THREAD

**Issue by nodiscc** [shaarli/Shaarli#102]
Firefox (Iceweasel) 34. Importing bookmarks using `Bookmark manager -> Import from HTML` does not import _all_ bookmarks but only a partial set. My bookmarks file has links and self-posts (notes), public and private. Firefox error console returns:

```
this._previousFrame is undefined BookmarkHTMLUtils.jsm:783
NS_ERROR_XPC_JAVASCRIPT_ERROR_WITH_DETAILS: Component returned failure code: 0x80570021 (NS_ERROR_XPC_JAVASCRIPT_ERROR_WITH_DETAILS) [nsINavBookmarksService.runInBatchMode] BookmarkHTMLUtils.jsm:934
Failed to import bookmarks from file:///home/gly/T%C3%A9l%C3%A9chargements/bookmarks_all_20150119_200026.html: [Exception... "Component returned failure code: 0x80570021 (NS_ERROR_XPC_JAVASCRIPT_ERROR_WITH_DETAILS) [nsINavBookmarksService.runInBatchMode]"  nsresult: "0x80570021 (NS_ERROR_XPC_JAVASCRIPT_ERROR_WITH_DETAILS)"  location: "JS frame :: resource://gre/modules/BookmarkHTMLUtils.jsm :: walkTreeForImport :: line 934"  data: no] BookmarkHTMLUtils.jsm:159
[Exception... "Component returned failure code: 0x80570021 (NS_ERROR_XPC_JAVASCRIPT_ERROR_WITH_DETAILS) [nsINavBookmarksService.runInBatchMode]"  nsresult: "0x80570021 (NS_ERROR_XPC_JAVASCRIPT_ERROR_WITH_DETAILS)"  location: "JS frame :: resource://gre/modules/BookmarkHTMLUtils.jsm :: walkTreeForImport :: line 934"  data: no]
```

**Comment by pikzen** [comment]
Could you post the .html file you're trying to import ? I'd like to test it on firefox 35.0. From what I'm seeing, this looks like a bug in Firefox (`this._previousFrame is undefined` points to an internal variable from Firefox's parser and apparently not the bookmark file, but this might be caused by a malformed bookmark export.). I have tried importing a small bookmark file (4 links) without any problems. 

```html
<!DOCTYPE NETSCAPE-Bookmark-file-1>
<!-- This is an automatically generated file.
     It will be read and overwritten.
     DO NOT EDIT! -->
<!-- Shaarli all bookmarks export on 2015/01/21 10:16:27 -->
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=UTF-8">
<TITLE>Bookmarks</TITLE>
<H1>Bookmarks</H1><DT><A HREF="http://reddit.com/r/programming" ADD_DATE="1421835336" PRIVATE="0" TAGS="tech,programming,software,opensource">Testing tests</A>
<DD>Tests testing tests testing tests in a test.
<DT><A HREF="http://news.ycombinator.com" ADD_DATE="1421835297" PRIVATE="0" TAGS="tech,programming,hipsters">HackerNews</A>
<DD>Bunch of hipsters
<DT><A HREF="http://sebsauvage.net/wiki/doku.php?id=php:shaarli" ADD_DATE="1316026800" PRIVATE="0" TAGS="opensource,software">Shaarli - sebsauvage.net</A>
<DD>Welcome to Shaarli ! This is a bookmark. To edit or delete me, you must first login.
<DT><A HREF="http://sebsauvage.net/paste/?8434b27936c09649#bR7XsXhoTiLcqCpQbmOpBi3rq2zzQUC5hBI7ZT1O3x8=" ADD_DATE="1315986322" PRIVATE="1" TAGS="secretstuff">My secret stuff... - Pastebin.com</A>
<DD>SShhhh!!  I'm a private link only YOU can see. You can delete me too.

```

**Comment by nodiscc** [comment]
Not reproducible anymore on the current `master`. Maybe I was on a dev branch when this happened.

**Comment by nodiscc** [comment]
Reopening because I still get an error when importing bookmarks using an export from a Shaarli on the current `master` branch. All bookmarks are not imported, only a partial set.

```
Failed to import bookmarks from file:///home/bsp/T%C3%A9l%C3%A9chargements/bookmarks_all_20150204_144401.html: [Exception... "Component returned failure code: 0x80570021 (NS_ERROR_XPC_JAVASCRIPT_ERROR_WITH_DETAILS) [nsINavBookmarksService.runInBatchMode]"  nsresult: "0x80570021 (NS_ERROR_XPC_JAVASCRIPT_ERROR_WITH_DETAILS)"  location: "JS frame :: resource://gre/modules/BookmarkHTMLUtils.jsm :: walkTreeForImport :: line 934"  data: no]
```

To investigate where the import started to fail, I imported all my public bookmarks (only links), which went well. I then tried to import my private bookmarks, which contain self-posts (Notes not liked to an URL). When the parser hits the first private _note_ (previous private _links_ import fine), it fails. This is the content of the note in the HTML export:

```
<DT><A HREF="?6SIQOA" ADD_DATE="1414176161" PRIVATE="1" TAGS="todo,video,list,index">Note: Old Vids</A>
<DD>Content-Type: text/x-zim-wiki
Wiki-Format: zim 0.4

Structure du dossier pour le volume QUARK500
Le num‚ro de s‚rie du volume est C2B3-8EFC
E:.

³   youtube_4lyUNs7eNhs-EscapeFromCity17PartTwo.webm
³   vimeo_34750078-Network.mp4
³   youtube_nTdpx_EZWnw-SooperLooperJackOSXTutorialPart1.webm
³   vimeo_38170231-AbandonedPlaces2YearsTimeLapse.mp4
³   youtube_hkDD03yeLnU-GUIinterfaceusingvisualbasictotrackthekillersIPaddressCSI.webm
³   vimeo_38610301-BoyCrisisLHommeOfficialMusic.mp4
³   vimeo_24030911-origins.mp4
³   youtube_PaMdXjTn9rc-CastleENHANCE.webm
³   vimeo_28922655-Synchronize.mp4
³   vimeo_38101676-TheMostAstoundingFact.mp4
³   ChumpAndClump.mp4
³   youtube_O2rGTXHvPCQ-Numb3rsdescriptionofIRC.webm
³   youtube_1uLSQ6QcQs4-BeautifulLibertyCityGrandTheftAutoIVMachinima.webm
³   LEDAHU.mp4
³   youtube_yVpbFMhOAwE-HowLinuxisBuilt.webm
³   vimeo_37830810-DreamMusicPart1.mp4
³   vimeo_38840688-unnamedsoundsculpture.mp4
³   vimeo_35330925-NewSumNoussommesHeyRosetta.mp4
³   youtube_Bo8Q88bAXyQ-scientistaircannonprank.webm
³   youtube_wgJfVRhotlQ-MinecraftScientificGraphingcalculatorSinCosTanLogSquareroot.webm
³   vimeo_38745437-SummerSolstice.mp4
³   vimeo_37101515-YelloBellyEnd.mp4
³   vimeo_37617460-Hazed.mp4
³   vimeo_27653079-TheTaleOfMrRˆvus.mp4
³   youtube_3EwZQddc3kY-EnemyofthestateFreezeRotateZoomFocusEnhance.webm
³   youtube_u8qgehH3kEQ-NCIS2IDIOTS1KEYBOARD.webm
³   chesse.log
³   youtube_9ixFf4ljuCg-EscapeFromCity17PartOne.webm
³   vimeo_38480783-HitchcockWalks.mp4
³   TheResistanceEpisode1WelcomeToAurodeca.mp4
³   youtube_J6ZWlDks0nQ-ColdplayParadiseOfficial.webm
³   MORTYS.mp4
³   youtube_KUFkb0d1kbU-RedDwarfBackToEarthPictureZoomSketch.flv
³   youtube_KA8a5qibUJU-ZeldaViolinDuetLindseyStirling.webm
³   youtube_76loVajVLZ4-CSIMiamienjoyingtheirreflection.flv
³   LIST
³   
ÀÄÄÄVimeo
```

**Comment by nodiscc** [comment]
I didn't look further into this, but my guess is that Firefox refuses to import this item because it has an invalid URL (

**Comment by alexisju** [comment]
I the same way, i think it's still a bug to import correctly from Delicious links + tags.

I fact, delicious support tags with space, who can't be imported corectly into Shaarli.

**Comment by nodiscc** [comment]
This should be reported to https://bugzilla.mozilla.org/show_bug.cgi?id=416611 as someone at Mozilla is working on import with tags. Let's ask if we can do something to improve compatibility, provided that:

 * We are still able to import from Delicious and legacy Shaarli exports
 * We are still able to store self-posts (`HREF="?6SIQOA`) in the export, and import them back.

**Comment by nodiscc** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=416611 has been solved and the patch is pending.

**Comment by mro** [comment]
wow, impressive bug history. @nodiscc thanks for posting this bug link. 7 years!

**Comment by nodiscc** [comment]
"Import with tags" seems to be fixed an merged in Firefox. The issue described on top of this issue thread is still present. We should investigate what Shaarli bookmark entries are causing this, and if possible find a workaround, or report the bug to Mozilla.

**Comment by ArthurHoaro** [comment]
Maybe @virtualtam current work on https://github.com/shaarli/netscape-bookmark-parser will fix you issue?

**Comment by virtualtam** [comment]
Importing Shaarli bookmarks in a web browser (or the other way around) may turn out a bit clumsy, for a couple reasons:
- the Netscape bookmark format is both basic and _very loose_ , in that it does not specify which kind of content it can handle or how it can be nested, AFAIK there's no DTD nor XSL to define:
    - which XML tags can be used
    - which XML attributes can be used
    - how data should be formatted (e.g. dates, tags)
- current Shaarli exports are not "clean" either, #538 will improve this, yet we're taking some freedom over some aspects
    - multiline descriptions
    - extra plugin formatting (Markdown)
    - mixed newlines styles depending on the OS & browser used to save a link (=> this may be worth sanitizing in `LinkDB`)

The current work on [netscape-bookmark-parser](https://github.com/shaarli/netscape-bookmark-parser) aims at improving a generic Netscape bookmark parser for scripting usage, i.e. importing bookmark dumps from miscellaneous sources to PHP, and potentially serve as a tool to help verifying the integrity of the bookmark dumps Shaarli generates.

Maybe we could improve compatibility with a given browser (importing Shaarli bookmarks) by generating bookmark files that are formatted _like_ said browser?

Feel free to open issues on:
- [Shaarli](https://github.com/shaarli/Shaarli/issues): Shaarli bookmark export compatibility with a given browser
- [shaarli/netscape-bookmark-parser](https://github.com/shaarli/netscape-bookmark-parser): Shaarli support for importing data from a browser dump
Bonus points if test data is provided, e.g. bookmark dumps that lead to faulty imports, or data that cannot be dumped properly.

**Comment by mro** [comment]
frankly I would recommend a stable & symmetric, lossless import/export in an existing web-standard plus conversion to and from the netscape files.

Candidates IMO are [OPML](https://en.wikipedia.org/wiki/Outline_Processor_Markup_Language) as well as [Atom / RFC4287](https://en.wikipedia.org/wiki/Atom_%28standard%29). Both have relaxng schemas ([OPML](https://github.com/mro/opml-schema/), [RFC4287](https://github.com/mro/iPhoto2Atom)), so they're fairly well defined.

Then Shaarli has to deal with a single well-defined format. The converters can focus on converting what makes development a lot easier and also makes them reusable and so maybe appealing to users and contributors outside the shaarli community.

P.S.: I guess you are aware the definition of the [netscape format over at Microsoft](https://msdn.microsoft.com/en-us/library/aa753582%28v=vs.85%29.aspx) but it didn't help your case.

**Comment by ArthurHoaro** [comment]
@mro We already can export everything in Atom format, but that doesn't help us regarding this netscape format.

> Maybe we could improve compatibility with a given browser (importing Shaarli bookmarks) by generating bookmark files that are formatted like said browser?

*If* we can't generate a file which will work in Firefox and Chrome, we can consider it.

**Comment by nodiscc** [comment]
> Maybe we could improve compatibility with a given browser
> generate a file which will work in Firefox and Chrome,

There are no reports of Shaarli bookmarks failing to import to Chrome yet (has it been tested?).  
There is however this bug about the Firefox import.  
I think ensuring compatibility with the FF import feature would be a great first step.

> bookmark dumps that lead to faulty imports, or data that cannot be dumped properly.

There are still no reports of Netscape Format bookmarks that fail to import in Shaarli.


> Bonus points if test data is provided, e.g. bookmark dumps that lead to faulty imports

I can work on this. It would probably help figuring out what causes my import to fail. What would be needed in the dump?

 * normal shaares (links), untagged
 * normal shaares (links), tagged
 * normal shaares (links), tagged with special characters
 * normal shaares (links), with single-line description
 * normal shaares (links), with multi-line description
 * normal shaares (links), with description containing HTML tags
 * self-posts
 * ... ?

**Comment by ArthurHoaro** [comment]
I made a few tests on the original issue with my database:

 * Firefox stops the import if an invalid URL is found (relative URL).
 * Chrome doesn't import invalid URL and goes on.

If I replace all relative URL with absolute ones, everything is imported in both browser, even long complicated markdown description. We just need to add a `index_url()` to relative URL in `NetscapeBookmarkUtils`.

**Comment by virtualtam** [comment]
@ArthurHoaro cool, I'll add this feature (prepending the Shaarli instance URL before note permalinks) in a future PR ;-)

@nodiscc mixed format shaares would be fine, the goal is to provide data sets that are representative of nominal usage, so tests can cover most usual cases

@mro I've mostly seen OPML used to handle feed data, e.g. importing/exporting feed subscriptions and bookmarks, and don't know to which extent it is supported by the different bookmarking applications/services (v.s. Netscape that is quite loose yet widely supported).
It also may not be built-in for web browsers without using extensions and/or feed aggregators (e.g. Firefox [OPML extension](https://addons.mozilla.org/en-US/firefox/addon/opml-support/))

**Comment by nodiscc** [comment]
Tested working, fixed by  #554, thanks to @virtualtam
